### DISCUSSION TITLE
Inquiry about Cropping Objects Using YOLOv8 Bounding Boxes and Saving Them

### DISCUSSION THREAD

**Discussion by AR97774-NipuniCabral** [ultralytics/ultralytics#10373] [Q&A]
Hello✨✨

I hope this message finds you well. I am currently working on an object detection project using YOLOv8 and have successfully trained my model. However, I am facing an issue when trying to crop objects from images using the bounding boxes generated by YOLOv8 and saving them into a designated folder.

I used the following code snippet to attempt this:

```
from ultralytics import YOLO
import cv2
import os

model = YOLO('/best.pt')
source = 'images (3).jpeg'
img = cv2.imread(source)

results = model.predict(source)

if results[0].boxes is not None:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img[y1:y2, x1:x2]
    
        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```

Unfortunately, I encountered an issue where objects were not detected when attempting to save the cropped images. The prediction without saving the crop works correctly and generates bounding boxes.

I would appreciate any guidance, insights, or code examples on how to resolve this issue and successfully crop objects using YOLOv8 bounding boxes. Thank you for your time and assistance.

Best regards.

**Comment by pderrenger** [comment]
Hello ✨,

I'm glad to hear about your progress with YOLOv8! Based on your description, it seems there might be a small mix-up in the way you’re fetching the image for cropping. The `model.predict()` function returns the results, but since you're providing a file path (`source`), you need to use the `results.render()` method to obtain the annotated image for cropping. However, since you're looking to crop the original detections, you should indeed be using the original image loaded with `cv2.imread(source)` as you have done.

The issue might be with how the `source` is being used in `model.predict(source)`. Since `source` is a filename, and you're also reading the image with `OpenCV`, ensure that the predictions and the original `cv2` image are correctly aligned. Let's modify your code slightly to check the results directly:

```python
from ultralytics import YOLO
import cv2
import os

model = YOLO('best.pt')  # Assuming your model file is named 'best.pt' and is in the current directory
source = 'images (3).jpeg'
img = cv2.imread(source)

# Directly use `img` for the prediction to ensure alignment
results = model.predict(img)

# Now you can process results[0].boxes.xyxy as before
if results[0].boxes is not None:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img[y1:y2, x1:x2]
    
        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```

This modification ensures you are cropping from the original image that corresponds exactly to the one used for prediction. If you're still facing issues, verify that your model's predictions are indeed generating bounding boxes by inspecting the `results[0].boxes` content more closely. Debugging tip: print the `box` values inside the loop to ensure they are within the expected range.

Hope this helps, and feel free to ask if you have further questions. Happy coding! 🚀

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨,

Thank you for your continued support and guidance regarding my YOLOv8 project. I greatly appreciate your assistance.

I followed your suggestions and made adjustments to my code as recommended. However, I encountered a persistent issue with cropping objects using the bounding boxes generated by YOLOv8. Despite the modifications, I still received the following output:

```
0: 448x640 505.8ms
Speed: 3.3ms preprocess, 505.8ms inference, 2.6ms postprocess per image at shape (1, 3, 448, 640)
No objects detected.
```
To provide more context and aid in resolving this issue, I will share the relevant part of my training codes:


```
pip install ultralytics
from IPython import display
display.clear_output()

import ultralytics
ultralytics.checks()
```

[output- Ultralytics YOLOv8.1.12 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 26.3/78.2 GB disk)]


```
%cd {HOME}
!yolo task=obb mode=train model=yolov8s-obb.pt data={dataset.location}/data.yaml epochs=50 imgsz=640 batch=16

```

Your advice on how to properly crop objects using the bounding boxes generated by YOLOv8 and save them to a designated folder would be immensely helpful.

Thank you once again for your constant support and assistance. I look forward to your guidance on resolving this issue.

**Reply by pderrenger** [reply]
Hello ✨,

Thank you for reaching out again, and I'm here to help! Based on the information provided, it seems like your model is not detecting any objects ("No objects detected."), which is why you’re unable to crop anything from the images.

The issue might not be with the cropping code but with the model's ability to detect objects in your images. It's important to ensure that the model has been correctly trained and is capable of detecting objects effectively. Here are a few things you might want to check:

- **Model Training and Evaluation:** Ensure your model is properly trained on your dataset. Review the training logs to see if the model is converging and achieving reasonable accuracy on your validation set. 
- **Detection Threshold:** Ensure the detection confidence threshold is not set too high, which might preclude detections. You can adjust it lower to see if you get detections, e.g., `conf=0.25` in your predict call.
- **Data Preprocessing:** Ensure the images you’re using for detection are preprocessed correctly (e.g., resized correctly as per your model's input size without altering the aspect ratio significantly).

If your model is detecting objects correctly in other contexts but not in your Python script, you might want to visually inspect the detections using `results.show()` before attempting to crop, just to confirm detections are happening:

```python
results = model.predict(img)
results.show() # Visually inspect detections
```

If detections are visually confirmed, and issues persist in cropping, a revisit to the cropping logic as per the previous instructions might be necessary. However, focus primarily on ensuring your model is detecting accurately.

Feel free to share any additional error messages or behaviors you observe. Your success with YOLOv8 is our top priority, and we're here to support you through this journey!

Best regards.

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨ 


The model performs well in terms of drawing bounding boxes when using the following prediction code:


```
from ultralytics import YOLO
import torch

model = YOLO('/content/drive/MyDrive/Research yolov8 obb epoch25/runs/obb/train/weights/best.pt')

# Define path to image file
source = '/content/i7.jpg'

# Run inference on the source
results = model.predict(source, show=True, save=True)
```

output

![676dc1e5-1256-4ed0-a1d9-732180d9db23](https://github.com/ultralytics/ultralytics/assets/100070791/fc62ac79-4d16-480c-8e42-724e50f40a59)




However, when I attempt to crop the detected objects using the following code:

```

from ultralytics import YOLO
import cv2
import os

model = YOLO('/content/drive/MyDrive/Research yolov8 obb epoch25/runs/obb/train/weights/best.pt') 
source = '/content/i7.jpg'
img = cv2.imread(source)

# Directly use `img` for the prediction to ensure alignment
results = model.predict(img)

if results[0].boxes is not None:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img[y1:y2, x1:x2]
    
        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```
The output consistently shows "No objects detected," despite successful detections in the previous code snippet. I have verified the model's accuracy with a validation set, achieving the following results:

- Class: all
- Images: 461
- Instances: 462
- Precision: 0.936
- Recall: 0.943
- mAP50: 0.975
- mAP50-95: 0.855

I would greatly appreciate any insights or suggestions you may have regarding this issue. Thank you for your assistance and support.

Best regards.

**Reply by pderrenger** [reply]
Hello ✨,

Thanks for providing detailed information about how you're using YOLOv8 and the issues you're facing with cropping detected objects! It looks like your model is performing well based on your validation metrics, which is great news.

The issue with the cropping script may be related to how the image is being read and used for prediction. When you use `cv2.imread(source)`, the image is loaded in BGR format, but YOLO expects the images in RGB format when provided as a NumPy array directly. To address this, you need to convert the image from BGR to RGB before passing it to the `predict()` function. Here's how you can modify your cropping code:

```python
from ultralytics import YOLO
import cv2
import os

model = YOLO('/content/drive/MyDrive/Research yolov8 obb epoch25/runs/obb/train/weights/best.pt')
source = '/content/i7.jpg'
img = cv2.imread(source)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)  # Convert from BGR to RGB

# Use the RGB image for prediction
results = model.predict(img)

if results[0].boxes:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        # Note: Convert back to BGR for saving correctly with OpenCV
        crop_img = cv2.cvtColor(img[y1:y2, x1:x2], cv2.COLOR_RGB2BGR)

        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```

This adjustment ensures that the image is in the correct format when predictions are made, which should hopefully solve the problem you're encountering. Please give this a try and let me know if it resolves the issue!

Best regards!

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨,

Thank you for your detailed response and continuous support regarding the object cropping issue.

I followed your recommendations and adjusted the code as suggested, converting the image from BGR to RGB before passing it to the predict() function. However, I still encountered the same issue of not detecting any objects for cropping. The output I received was:

```

: 480x640 465.7ms
Speed: 3.8ms preprocess, 465.7ms inference, 1.8ms postprocess per image at shape (1, 3, 480, 640)
No objects detected.
```
Upon further investigation, I found that this problem is not unique to my setup but has been reported by multiple users. One user shared the following output as an example of the issue:

```

masks: None
names: {0: 'Common Krait', 1: 'Indian Cobra', 2: 'Green Vine', 3: 'Python', 4: 'Viper', 5: 'pit viper',}
obb: ultralytics.engine.results.OBB object
orig_img: array([[[215, 215, 215],
[215, 215, 215],
[215, 215, 215],
...,
[225, 225, 225],
[226, 226, 226],
[226, 226, 226]],

[[216, 216, 216],
[216, 216, 216],
[216, 216, 216],
...,
[223, 223, 223],
[224, 224, 224],
[224, 224, 224]],

[[217, 217, 217],
[217, 217, 217],
[217, 217, 217],
...,
[221, 221, 221],
[221, 221, 221],
[221, 221, 221]],

...,

[[107, 107, 107],
[106, 106, 106],
[103, 103, 103],
...,
[ 88,  88,  88],
[ 89,  89,  89],
[ 86,  86,  86]],

[[115, 115, 115],
[113, 113, 113],
[107, 107, 107],
...,
[ 87,  87,  87],
[ 90,  90,  90],
[ 89,  89,  89]],

[[120, 120, 120],
[116, 116, 116],
[109, 109, 109],
...,
[ 88,  88,  88],
[ 94,  94,  94],
[ 94,  94,  94]]], dtype=uint8)
orig_shape: (1600, 1200)
path: '/content/grayscale_image3.jpeg'
probs: None
save_dir: 'runs/obb/predict'
speed: {'preprocess': 7.385492324829102, 'inference': 3030.5862426757812, 'postprocess': 15.470743179321289}
```
This issue is causing confusion as the model can detect objects correctly, but the results[0] object does not show valid probs and boxes. Your assistance in resolving this matter would be greatly appreciated.

Thank you once again for your support.

Best regards.

**Reply by pderrenger** [reply]
Hello ✨,

Thank you for your update and for sharing additional details about the issue you're facing. I'm sorry to hear that the problem persists.

Based on the new information, it seems there might be an issue with how the results are being processed or a specific scenario where the model behaves unexpectedly. To help diagnose further, could you please verify the following in your environment?

1. Ensure the model weights are loaded correctly and the same model configuration is used during both training and prediction.
2. Check if the issue persists with different images or just specific ones.
3. Attempt to log the confidence thresholds and other related parameters during prediction to ensure they aren't set too strictly.

If the problem continues, it might be helpful to isolate whether the preprocessing or model itself is behaving unexpectedly. Testing with known detection-friendly images could help establish a baseline performance.

Here’s a quick snippet to log additional details during predictions:
```python
results = model.predict(img, conf=0.25)  # Adjust confidence threshold
print("Detected boxes:", results[0].boxes)
print("Model confidence thresholds:", model.conf)
```
This may provide more insights into what may be causing the discrepancy. Looking forward to your observations, and let's continue working towards a resolution!

Best regards.

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨,

Thank you for your suggestions and ongoing support regarding the object cropping issue.

I'd like to provide some updates based on your recommendations:

### Training Code Used:
The training code I used is as follows:

`!yolo task=obb mode=train model=yolov8s-obb.pt data='/content/drive/MyDrive/yolo/datasets/Snake-Research-1/data.yaml' epochs=5 imgsz=640 batch=16`

### Testing with Different Images:
I tried running inference using different images, and each time the model accurately drew bounding boxes around the detected objects. However, when attempting to crop the objects using the bounding boxes, I consistently received the message "no object detected."

### Log of Confidence Thresholds and Parameters:

I ran the code snippet you provided to log additional details during predictions, with a confidence threshold of 0.25:
```

results = model.predict(img, conf=0.25)
print("Detected boxes:", results[0].boxes)
print("Model confidence thresholds:", model.conf)
```
Here is the output I received:


```
image 1/1 /content/i7.jpg: 416x640 352.9ms
Speed: 3.1ms preprocess, 352.9ms inference, 1.5ms postprocess per image at shape (1, 3, 416, 640)
Results saved to runs/obb/predict5
Detected boxes: None
Model confidence thresholds: 0.25
```

Result Image:
![676dc1e5-1256-4ed0-a1d9-732180d9db23](https://github.com/ultralytics/ultralytics/assets/100070791/36c3a146-7fe7-4d45-b9eb-f16bbb0068c6)



Please update me if I have overlooked something or if there are any specific steps I should follow to resolve this issue. Your continued support is greatly appreciated.

Best regards.

**Reply by pderrenger** [reply]
Hello ✨,

Thank you for your detailed update and for trying out the suggested steps. It's great to hear that your model accurately recognizes and frames the objects with bounding boxes.

The discrepancy you're encountering—bounding boxes being detected but then "no object detected" when attempting to crop—might suggest a possible inconsistency or bug in how bounding box objects are being accessed or reported post-detection.

Here’s an efficiently concise approach to try:

- Ensure that the bounding box data is being correctly parsed in your script. Sometimes direct access methods like `results[0].boxes.xyxy` are overlooked in favor of simpler `results[0].boxes` which might not directly translate to usable coordinates in every context.

- You might also check whether the bounding boxes are within the valid dimensions/range of the source image before attempting to crop. This might be causing an issue if the crop dimensions fall outside the actual image dimensions.

Here's a quick verification script:

```python
if results[0].boxes:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        if x1 < img.shape[1] and x2 < img.shape[1] and y1 < img.shape[0] and y2 < img.shape[0]:
            crop_img = img[y1:y2, x1:x2]
            cv2.imwrite(os.path.join('cropped_images', f'crop_{i}.jpg'), crop_img)
        else:
            print(f"Skipping box {i}, dimensions out of bound.")
else:
    print("No objects with valid bounding boxes detected.")
```

This script ensures only crop attempts within the valid image boundaries to avoid potential errors in cropping or image handling.

Looking forward to your next update and we're here to help every step of the way! 🚀

Best regards.

**Reply by cabral0413** [reply]
Dear YOLOv8 Team,

I am working on a snake classification project and using YOLOv8 for both object detection and classification tasks. I have encountered a couple of questions and would appreciate your guidance.

Cropping and Classifying Images:

I am currently using YOLOv8 for object detection with the following code snippet to crop images:

```
from ultralytics import YOLO
import torch

model = YOLO('/content/drive/MyDrive/YolodefFinal/runs/detect/train3/weights/best.pt')

# Define path to video file
source = '/content/GV.jpeg'

# Run inference on the source
results = model.predict(source, show=True, save=True, save_crop=True)
```
This code successfully crops images. However, instead of saving these cropped images, I want to pass them directly to the classification model within the same script.

Current Classification Process:
In the classification phase, I use the following code to perform classification:


```
from flask import Flask, request
from ultralytics import YOLO
from PIL import Image

app = Flask(__name__)

@app.route('/classify', methods=['POST'])
def predict():
    file = request.files['image']
    img = Image.open(file.stream).convert("RGB")  # Ensure it's in RGB format
    
    model = YOLO('models/best.pt')  # Make sure the path is correct
    results = model(img)  # Directly pass PIL image

    # Check if any detections are made
    if len(results) == 0:
        return {'message': 'No objects detected in the image.'}
    
    # Assuming we're doing classification, let's find the top two classes if available
    if len(results) >= 2:
        top_classes = [(results[i].names[results[i].probs.top1], results[i].probs.top1conf.item()) for i in range(2)]
    else:
        top_classes = [(results[i].names[results[i].probs.top1], results[i].probs.top1conf.item()) for i in range(len(results))]
    
    # Determine venom status for each top class
    top_venom_status = [get_venom_status(class_name) for class_name, _ in top_classes]
    
    # Format the probabilities as percentages with two decimal points
    formatted_probs = ["{:.2%}".format(confidence) for _, confidence in top_classes]

    # Check if confidence is below 70% for any top class
    if any(confidence < 0.7 for _, confidence in top_classes):
        return {'message': 'No snakes detected with sufficient confidence. Please upload a clearer photo.'}

    return {'top_predictions': [{'class': class_name, 'probability': prob, 'venom_status': venom_status}
                                for (class_name, _), prob, venom_status in zip(top_classes, formatted_probs, top_venom_status)]}

def get_venom_status(class_name):
    # Implement your logic to determine venom status based on the class name
    if class_name == 'Common Indian Krait':
        return 'Venomous'
    elif class_name == 'Python':
        return 'Venomous'
    elif class_name == 'Hump Nosed Viper':
        return 'Venomous'
    elif class_name == 'Green Vine Snake':
        return 'Non-venomous'
    elif class_name == 'Russells Viper':
        return 'Venomous'
    elif class_name == 'Indian Cobra':
        return 'Venomous'
    else:
        return 'Unknown'

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000, debug=True)
```

Goal:
I want to modify my app.py file to first crop the image using object detection and then send it directly to the classification model to classify and show the results. I have tried to integrate these two parts but encountered errors.

Could you please provide guidance on how to efficiently combine the object detection and classification processes within the same script?

Thank you for your assistance.

Best regards,
Nipuni Cabral
